### ISSUE TITLE
Unable to select first character between two punctuation marks in Chinese epub file

### ISSUE THREAD

**Issue by fromhz** [koreader/koreader#6268]
* KOReader version: KOReader-v2020.04.1.zip
* Device: Kobo Forma
(installed with kfmon)

#### Issue
Unable to select first character between two punctuation marks in Chinese epub file

#### Steps to reproduce
Open a Chinese book in koreader
Attempt to select first character between two punctuation marks (or right after a line break and before a punctuation mark)
example text
`一二三四。`
The entire phrase between the two punctuation marks are selected.
e.g.
`一二三四` are selected instead of just `一`

Impact: It is okay if there are only three characters or less in the phrase, koreader gives me dictionary outputs for each character. When there are four or more, however, koreader tries to find definition of the entire phrase in the dictionary, which of course fails.

Thank you!

**Comment by poire-z** [comment]
OK, I can reproduce the issue.
Just putting some debugging notes, because I'm not familiar with this part of the code, and I can't investigate that much right now:

<kbd>![image](https://user-images.githubusercontent.com/24273478/84589753-1a886b80-ae31-11ea-9c97-b8655e5cff32.png)</kbd>

cre.cpp:
```c
static int getTextFromPositions(lua_State *L) {
[...]
// When panning, only extend to include the whole word when we are
// actually holding inside a word. This allows selecting punctuation,
// quotes or parens at start or end of selection to have them
// included in the highlight.
bool not_panning = r.getStart() == r.getEnd();
printf("(not_panning %d || r.getStart().isVisibleWordChar() %d ) && !r.getStart().isVisibleWordStart() %d\n", not_panning, r.getStart().isVisibleWordChar(), !r.getStart().isVisibleWordStart());
printf("(not_panning %d || r.getEnd().isVisibleWordChar() %d) && !r.getEnd().isVisibleWordEnd() %d\n", not_panning, r.getEnd().isVisibleWordChar(), !r.getEnd().isVisibleWordEnd());;
if ( (not_panning || r.getStart().isVisibleWordChar()) && !r.getStart().isVisibleWordStart())
        r.getStart().prevVisibleWordStart();
if ( (not_panning || r.getEnd().isVisibleWordChar()) && !r.getEnd().isVisibleWordEnd())
        r.getEnd().nextVisibleWordEnd();
```

crengine/src/lvtinydom.cpp
```c
/// returns true if current position is visible word end
bool ldomXPointerEx::isVisibleWordEnd()
{
    if ( isNull() )
        return false;
    if ( !isText() || !isVisible() )
        return false;
    ldomNode * node = getNode();
    lString16 text = node->getText();
    int textLen = text.length();
    int i = _data->getOffset();
    lChar16 currCh = i>0 ? text[i-1] : 0;
    lChar16 nextCh = i<textLen ? text[i] : 0;
    printf("i %d textLen %d, canWrapWordAfter(currCh) %d || (!IsWordSeparator(currCh) %d && IsWordSeparatorOrNull(nextCh) %d )\n", i, textLen, canWrapWordAfter(currCh), !IsWordSeparator(currCh), IsWordSeparatorOrNull(nextCh));
    if (canWrapWordAfter(currCh) || (!IsWordSeparator(currCh) && IsWordSeparatorOrNull(nextCh)))
        return true;
    return false;
}
```
When selecting 2nd, 4th, and then 1st glyph:
```
2nd:
(not_panning 1 || r.getStart().isVisibleWordChar() 1 ) && !r.getStart().isVisibleWordStart() 0
i 1 textLen 26, canWrapWordAfter(currCh) 1 || (!IsWordSeparator(currCh) 1 && IsWordSeparatorOrNull(nextCh) 0 )
(not_panning 1 || r.getEnd().isVisibleWordChar() 1) && !r.getEnd().isVisibleWordEnd() 0
i 1 textLen 26, canWrapWordAfter(currCh) 1 || (!IsWordSeparator(currCh) 1 && IsWordSeparatorOrNull(nextCh) 0 )

4th:
(not_panning 1 || r.getStart().isVisibleWordChar() 1 ) && !r.getStart().isVisibleWordStart() 0
i 3 textLen 26, canWrapWordAfter(currCh) 1 || (!IsWordSeparator(currCh) 1 && IsWordSeparatorOrNull(nextCh) 0 )
(not_panning 1 || r.getEnd().isVisibleWordChar() 1) && !r.getEnd().isVisibleWordEnd() 0
i 3 textLen 26, canWrapWordAfter(currCh) 1 || (!IsWordSeparator(currCh) 1 && IsWordSeparatorOrNull(nextCh) 0 )

1st:
(not_panning 1 || r.getStart().isVisibleWordChar() 1 ) && !r.getStart().isVisibleWordStart() 0
i 0 textLen 26, canWrapWordAfter(currCh) 0 || (!IsWordSeparator(currCh) 0 && IsWordSeparatorOrNull(nextCh) 0 )
(not_panning 1 || r.getEnd().isVisibleWordChar() 1) && !r.getEnd().isVisibleWordEnd() 1
i 0 textLen 26, canWrapWordAfter(currCh) 0 || (!IsWordSeparator(currCh) 0 && IsWordSeparatorOrNull(nextCh) 0 )
```

There's something with `lChar16 currCh = i>0 ? text[i-1] : 0;` : when 1st glyph (i=0), we're testing glyph 0 instead of the `text[0]`.
So, may be some special case to handle there when i=0, or there's something more generally off by one (which could be, with english text, we wrap after the space following a word).

**Comment by poire-z** [comment]
Btw, found a trick while testing.
When long-pressing on the first glyph:
<kbd>![image](https://user-images.githubusercontent.com/24273478/84589906-58d25a80-ae32-11ea-8dce-eb9dd9f3a2e8.png)</kbd>
Just move your fingers on the right (while still pressing) as if you wanted to include the 2nd glyph in selection, and I magically get:
<kbd>![image](https://user-images.githubusercontent.com/24273478/84589922-7b647380-ae32-11ea-9334-2f0b528bf886.png)</kbd>
(well, this works only if you have a font large enough so that moving half other 2nd glyph triggers getting over or pan threshold.)

Regarding your comment in https://github.com/koreader/koreader/issues/6162#issuecomment-643729818 :
> N: It would be great if we can readjust highlighting with the floating cursor balls (like in kobo nickel system). It is really hard to get the selection right in SC compared to English, and for dictionary lookup, sometimes I need to look up a word consisting of multiple characters. It is nearly impossible to do so correctly the first time.

You can tweak the selection, and also solve your issue above of selecting the first glyph.
Anytime, just panning as suggested above (and so, usually having more than one glyph selected) will bring the highlight dialog instead of launching a dict lookup:
<kbd>![image](https://user-images.githubusercontent.com/24273478/84590002-0d6c7c00-ae33-11ea-82ac-999729565e07.png)</kbd>

Just select `Highlight`.
Then, Tap on the highlighted text, which brings the highlight editing popup:
<kbd>![image](https://user-images.githubusercontent.com/24273478/84590017-2f65fe80-ae33-11ea-9412-956972cfc913.png)</kbd>

Short tap on the arrow will move by western words (up to some space or punctuation), but long-press on the arrow will move by glyph. So, 2 long press on that yellow arrow will allow you to get only the 1st glyph selected:
<kbd>![image](https://user-images.githubusercontent.com/24273478/84590057-88ce2d80-ae33-11ea-86b1-177f4d9158ab.png)</kbd>

Then, hitting the ellipsis button (in blue) will bring you back the 1st dialog, where you can launch a dict lookup:
<kbd>![image](https://user-images.githubusercontent.com/24273478/84590079-c0d57080-ae33-11ea-9bf8-b2e4b626b87e.png)</kbd>

**Comment by fromhz** [comment]
Thanks for your reply XD
> (well, this works only if you have a font large enough so that moving half other 2nd glyph triggers getting over our pan threshold.)

There are 11 font size in the AA menu, let's call them 1 (smallest)~11 (largest). I successfully used your strategy with size 9 and above, but I find size around 3 to be a more reasonable choice for reading (on my device at least).

> You can tweak the selection, and also solve your issue above of selecting the first glyph.

Wow, this is actually a really handy feature for highlighting! Might be a bit too much work for dictionary look up though, since sometimes there are ~10 characters between two punctuations.

**Comment by poire-z** [comment]
OK, this should be solved by https://github.com/koreader/crengine/pull/347#issuecomment-644732023.
(Probably won't be bumped into frontend until middle of next week.)

**Comment by fromhz** [comment]
Thanks XD
I'll check it out when it is :)

**Comment by poire-z** [comment]
Hopefully fixed by #6305 , in tomorrow's nightly.
